### DISCUSSION TITLE
Problem with new install

### DISCUSSION THREAD

**Discussion by heffneil** [jgyates/genmon#1170] [Questions] [answered]
### Checklist

- [X] I made sure that there are *no existing issues or discussions* - [open issues](https://github.com/jgyates/genmon/issues),[closed issues](https://github.com/jgyates/genmon/issues?q=is%3Aissue+is%3Aclosed),[open discussion](https://github.com/jgyates/genmon/discussions), [closed discussion](https://github.com/jgyates/genmon/discussions/categories/questions?discussions_q=is%3Aclosed) - which may answer my question or that I could contribute my information to.
- [X] I have read the [Known Issues Wiki Page](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues) and my question / issue is not listed.
- [X] I have provided the version of the software and the operating system in my post.
- [X] If I have an issue with genmon not communicating with the controller, I have read and followed the [serial troubleshooting guide](https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting).
- [X] I have included the generator make and model in my submission.

### Version

48KW Liquid Cooled Generac with a Evolution controller. 
Pihat with Raspberry Pi 4.
Powered by battery on the generator with the pihat.

### Have you submitted your log files from the About page? Note that outbound email must be working for logs to be submitted.

I have not submitted logs

### What is/are your question(s)?

Hey guys I have a 48KW Liquid Cooled Generac with a Evolution controller.  I currently have it all disassembled right now because to get to the plugs they are behind everything.  I have extended the included molex cable with the pintsize hat.  I have the software running but I don't think I am getting communications from the system.  When I purchased the hat I had selected: 
Form: Standard v2.3
Generator Controller: Evolution 2 Liquid Cooled

The problem is I have a Evolution (1) I don't know if that matters or is causing my issues.  I am trying to button this up and the sun starts to bake on that side of the house.  Any recommendations on how I could have screwed this up would be useful.
I am not sure if the default /dev/serial0 is correct for the hat?

**Comment by skipfire** [comment]
The hardware is the same, the purchase process asks about the controller just so that if there is a support request the controller information has been provided. /dev/serial0 should be correct for all Raspberry Pi's before the Pi 5.

The V2.3 boards haven't been sold for quite awhile, did this ever work or are you just now getting around to installing it?

What genmon version and Pi OS are you using per checkbox 3?

What were the results of the serial troubleshooting you did per checkbox 4?

You said you don't think you are getting communications from the system, what symptoms are you having?

Do you have outbound email setup so that you can send the logs?

**Comment by heffneil** [comment]
I am just getting around to installing it.  Quite a while I think I bought it less than a year ago? Looks like September 23 so just under!

I can't seem to get any information from the evolution.  I don't have any data from the generator and I am getting Timeout Errors:

```Communication Stats :
        Packet Count : M: 125, S: 0
        CRC Errors : 0
        CRC Percent Errors : 0.00%
        Timeout Errors : 124
        Timeout Percent Errors : 99.20%
        Modbus Exceptions : 0
        Validation Errors : 0
        Sync Errors : 0
        Invalid Data : 0
        Discarded Bytes : 0
        Comm Restarts : 0
        Packets Per Second : 0.33
        Modbus Transport : Serial
        Serial Data Rate : 9600

```


```sudo python3 serialconfig.py -c

Using files: /boot/firmware/config.txt and /boot/firmware/cmdline.txt

Model: Raspberry Pi 4 Model B Rev 1.5

Checking : Is enable UART in boot config: OK
Checking : Is BT overlay disabled: OK
Checking : Serial console command line removed: OK
Checking : Serial console service disabled: OK
Checking : BT service disabled: OK```

**Reply by heffneil** [reply]
I do have email setup and working.  When I unplugged the cable that was on the back of the unit originally there were only 2 wires used.  I extended those 2 wires (rs485 send and receive on the hat instructions) so I can locate the pi elsewhere.  No other wires are extended and I don't believe required - but not sure.  I am not using the splitter I also purchased for the generators wifi and the pi - I want to get the PI working properly first.

I just installed Genmon 2 days ago so I have to belive this is the latest version from the git.
Genmon
Version V1.19.04

**Reply by skipfire** [reply]
Did it work without the extension? 

The typical minimum required is both data lines and a ground. As for the extension itself I'm not sure what instructions you're referring to as I don't provide instructions on how to do extenders for the HATs, but genman requires RS232 not RS485 so that and The lack of an extended ground could be your problem.

**Comment by heffneil** [comment]
Submitted logs on the About page.

**Comment by heffneil** [comment]
The instructions for the hat show the pin out for the molex plug and it shows RS485.  So I cut those two wires - which are the only two used in the factory plug and extended them straight away.  I did not test before extending so I can't really say if it worked. I will go check with a meter that the pins are completely connected since I guess its possible. 

My generator is in the stopped situation because I had to unplug the one safety switch to get the outside case off but I would think I get data still like battery voltage?

**Reply by skipfire** [reply]
The generac mobile link uses RS485, but other pins are used by genmon for RS232 so if you're using the generac cable as a template it will not work.

**Reply by jgyates** [reply]
Evo1 does not use RS-485, only RS232. this page has the pinout for the molex: https://github.com/jgyates/genmon/wiki/3.1--Making-a-Cable

to add to the point @skipfire is making, the Evo2 communicates with the new wifi mobile link adapter via RS-485. the Evo1 does not have support for RS-485. You may have the orientation reference wrong on the pinout. You need 3 connections for Evo1 on the molex: Transmit, Receive and Ground. If you only have 2 serial comms will not work.

**Reply by heffneil** [reply]
I followed the cable that was plugged in to the accessory port and I swear it looks like its just the antenna?! Is that possible on this generator?

**Reply by skipfire** [reply]
The module that would have been plugged in there is MobileLink and it is a pretty small device, mine is about the size of a tuna can, I believe some others have small boxes with the antenna on them as well, without seeing it is hard to be certain, but it sounds like the external antenna style of MobileLink.

**Reply by jgyates** [reply]
Maybe a picture may help. you can attach pictures via the web interface of github by dropping them into the text input area in this thread. I am not sure what you mean by antenna.

**Reply by heffneil** [reply]
> The module that would have been plugged in there is MobileLink and it is a pretty small device, mine is about the size of a tuna can, I believe some others have small boxes with the antenna on them as well, without seeing it is hard to be certain, but it sounds like the external antenna style of MobileLink.

Mine isn't anything like this.  My controller is huge.  The two wires appear to go up to the external antenna area and there seems to be another plug to another device.  I will go try to get more pics.

**Reply by heffneil** [reply]
> The module that would have been plugged in there is MobileLink and it is a pretty small device, mine is about the size of a tuna can, I believe some others have small boxes with the antenna on them as well, without seeing it is hard to be certain, but it sounds like the external antenna style of MobileLink.

Ok I found that box it is behind the antenna.  I couldn't see it before but you are right it is there.

**Comment by heffneil** [comment]
I received all this equipment a year ago and I’m just installing it now some of it may have gotten mixed up. I have this Molex cable that was in the box with the hat and now that you have me checking resistance I noticed that the two pins that are blue and white on this cable on one end of the opposite on the other end. I guess the question is should this cable be straight through or does it have to flip between the evolution and the hat?

**Reply by skipfire** [reply]
When unmodified the cable can be flipped around.  The ends do mirror left to right so pins 1 2 3 4 on one end are pins 4 3 2 1 on the other end. This happens with the power pins as well, so you have to be very careful not to end up shorting the system out by hooking up 12 volt battery to generator ground with the modified cable.

**Reply by heffneil** [reply]
I am sorry I don't understand - I am powering from the barrel connector on the hat.  Now I have the cable where I just connected the same cable to where it was originally through on the 2 pins (blue and white) I tried it with the other extended cable and it didn't work so I am completely lost on where I have a problem.

**Comment by heffneil** [comment]
I just found another cable in the box with my other stuff and it’s part number 451350810

**Reply by skipfire** [reply]
That is the 1 meter cable.  The cables all have the same pinouts, just different lengths.

**Reply by heffneil** [reply]
This cable from the pi hat to the controller should work, right?

**Reply by skipfire** [reply]
Yes, the cables are all the same pinouts, just different lengths.

**Reply by heffneil** [reply]
Ok well that cable from the accessory port to my pi hat is not working.  Any suggestions on what to test please?  I am very lost and I don't know what could be the issue here now.

**Reply by skipfire** [reply]
At this point I believe you have the cable from Accessory Port 1 to the HAT and the power cable from the battery to the HAT, is that correct?

Not working covers a ton of possibilities.  Is the Pi not powering up?  Is it not accessible on the network?  Is GenMon not showing that it is connected? What were the results of the serial testing you performed for checkbox 4?

**Reply by heffneil** [reply]
Yes this is the correct connection situation.  I can reach genmon - and I am getting read errors as stated above.  I can't seem to connect and get data like battery voltage.  I ran the serial connection script and the verify and posted the results above?

**Reply by skipfire** [reply]
What about the results from serialtest.py?

**Reply by jgyates** [reply]
I received a set of logs today at 12:39 EST, I assume this was from @heffneil. The serial results are:

Comm Stats :
   Packet Count : M: 14, S: 0
   CRC Errors : 0
   CRC Percent Errors : 0.00%
   Timeout Errors : 13
   Timeout Percent Errors : 92.86%
   Modbus Exceptions : 0
   Validation Errors : 0
   Sync Errors : 0
   Invalid Data : 0
   Discarded Bytes : 0
   Comm Restarts : 0
   Packets Per Second : 0.35
   Modbus Transport : Serial
   Serial Data Rate : 9600

Timeout errors indicate the pi thinks it is transmitting, but nothing is getting back. This is typically either cabling or power/ground issues.

**Reply by heffneil** [reply]
Right that is mine.  I am grounded via the battery, right?



Here are a few pics not sure they will be informative

https://imgur.com/a/WwinwOU

The other port above the accessory is just tape there isn't a plug on the back under that tape.

``` Using files: /boot/firmware/config.txt and /boot/firmware/cmdline.txt

Model: Raspberry Pi 4 Model B Rev 1.5

Checking : Is enable UART in boot config: OK
Checking : Is BT overlay disabled: OK
Checking : Serial console command line removed: OK
Checking : Serial console service disabled: OK
Checking : BT service disabled: OK

**Reply by skipfire** [reply]
That image looks like the right plug given the other pictures I have seen for Evo2 LC.

I would rely on grounding with the Molex cable not just the battery cable.  The reverse polarity protection on the barrel plug does shift the voltage slightly and without the ground on the Molex that "might" throw it off just enough to not work.

**Reply by heffneil** [reply]
Ok but I am using the regular cable from the pi to the controller and it still failed to get data?

**Reply by skipfire** [reply]
I would suggest finishing the documented serial troubleshooting as the serialtest.py step will confirm if the problem is in the Pi OS/genmon/HAT, or if it is outside that assembly.

**Reply by jgyates** [reply]
Here is another thread with some Evo1 Liquid Cooled picks. https://github.com/jgyates/genmon/discussions/1073

I am a little confused by the picture you sent. It looks like you have an Evo 1, but you also have a splitter from pintsize.me with something with 2 wires connected to it. Is this wifi mobile link or LTE? Was mobile link working with the generator before?

I was not aware that the mobile link adapter used RS-485 on Evo1. If it does not then the splitter will not work. Is that correct @skipfire?

I don't see a picture with everything connected as you are intending to use it. That may clarify thing for me.

**Reply by skipfire** [reply]
You are correct @jgyates, on second review I do see that the MobileLink was on the left 2 pins which are the RS232 pins, that leaves either MobileLink or GenMon, but not both.

**Reply by heffneil** [reply]
Yes I am wrong I thought I had an Evo2 when I ordered.  I need to remove the splitter and go straight through but I still have the problem of NOT getting data with the pre-prepared long cable?

**Comment by heffneil** [comment]
Ok here are more pics and maybe more informative.  I REALLY don't think I have a wiring problem but what do I know?

https://imgur.com/a/W9p4VKM

i plugged the mobile link back in and was able to get it to work although it does have a stop condition.  Currently NOW I have mobile link still working with the splitter so I think I have more progress.  

This now comes to the question of which wires need to be extended beyond the two in the top left corner?

It sounds like this is rs232 so the splitter will not work properly am I correct about that?  

Thanks for all the help guys I much appreciate it.

**Reply by skipfire** [reply]
You are in an either/or situation with GenMon or MobileLink.  I would remove both the splitter and the spliced cable and see what happens, if that doesn't work then you really need to do the serialtest.py to determine if it is OS/genmon/HAT side or if it is cable/generator side.  If the OS/genmon/HAT tests ok, you can even add the cable into the mix and retry the serialtest.py and if it still works the only thing not covered is the controller.

**Reply by jgyates** [reply]
since you have an evo1, yes. You can not use mobile link and genmon at the same time since they use the same serial port.

**Reply by jgyates** [reply]
what happens if you remove the splitter and plug the unmodified cable directly from the controller to the HAT?

**Reply by heffneil** [reply]
> what happens if you remove the splitter and plug the unmodified cable directly from the controller to the HAT?

Right I didn't have success with this.  I would assume it would start working rather immediately.  Im going back out to remove the splitter.

**Reply by jgyates** [reply]
once the splitter is removed, let's see if you get the same type of serial errors in the web interface (timeout). If you do then we can do a few tests to determine if your problem is the controller, the HAT, the PI or the cable. The steps to determine this are outlined on this page: https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting

Let us know if you have questions.

**Comment by heffneil** [comment]
```
Monitor :

    Generator Monitor Stats :
        Monitor Health : System Initializing. Not receiving data.
        Controller : Nexus, Air Cooled
        Run time : Generator Monitor running for 0:43:44.
        Generator Monitor Version : V1.19.04
        Update Available : No


    Communication Stats :
        Packet Count : M: 868, S: 0
        CRC Errors : 0
        CRC Percent Errors : 0.00%
        Timeout Errors : 867
        Timeout Percent Errors : 99.88%
        Modbus Exceptions : 0
        Validation Errors : 0
        Sync Errors : 0
        Invalid Data : 0
        Discarded Bytes : 0
        Comm Restarts : 0
        Packets Per Second : 0.33
        Modbus Transport : Serial
        Serial Data Rate : 9600


    Platform Stats :
        CPU Temperature : 138.92 F
        Pi Model : Raspberry Pi 4 Model B Rev 1.5
        Pi CPU Frequency Throttling : OK
        Pi ARM Frequency Cap : OK
        Pi Undervoltage : OK
        CPU Utilization : 1.3%
        OS Name : Debian GNU/Linux
        OS Version : 12 (bookworm)
        System Uptime : 0:44:22
        Network Interface Used : wlan0
        WLAN Signal Level : -69 dBm
        WLAN Signal Quality : 41/70
        WLAN Signal Noise : -256 dBm
        WLAN ESSID : IOTDevices
        System Time : Saturday August 24, 2024 15:33:26```

**Reply by jgyates** [reply]
looks like more timeout errors. 

The serial loopback tests are described [here](https://github.com/jgyates/genmon/wiki/3.6---Serial-Troubleshooting), but the short version is there are three tests.

1) loopback at the pi between pins on the pi (without the HAT connected) and run serialtest.py. This validates the pi is working a setup right
2) loopback at the HAT connector. This validate the HAT is functional
3) looback at the far end of the cable with it connected to the HAT. This validates the cable is functional

Also, since you are running OS version 12 (bookworm) you are in a managed python environment so use the commands on [this page](https://github.com/jgyates/genmon/wiki/Appendix-S---Working-in-a-Managed-Python-Environment) to run serialtest.py

**Comment by heffneil** [comment]
Good news I think:

```
(genenv) generac@generac:~/genmon $ sudo ./genenv/bin/python3 ./OtherApps/serialtest.py

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5
Serial port opened
write data: sent test string
waiting to received data....
error communicating...: 'utf-8' codec can't decode byte 0x99 in position 2: invalid start byte serialtest.py:177
(genenv) generac@generac:~/genmon $

**Reply by heffneil** [reply]
OK disregard my happiness I forgot to stop genmon:
```
(genenv) generac@generac:~/genmon $ sudo ./genenv/bin/python3 ./OtherApps/serialtest.py

Note: Genmon must NOT be running for this test to work properly. If Genmon is running this test will not function properly,

Loopback testing for serial port /dev/serial0...

Using python serial library V3.5
Serial port opened
write data: sent test string
waiting to received data....
PASSED! Loopback successful

**Reply by skipfire** [reply]
That narrows it to cable or controller, it indicates the Pi/OS/HAT are all fine.

**Reply by heffneil** [reply]
Well how do we know the hat is ok?  Which pins do I loop to test the hat?  Is there any switches on the hat that could cause this peculiar behavior?  When I went to put the hat back on I was convinced I previous misaligned the pins but unfortunately that's not the case :(

**Reply by skipfire** [reply]
You did the loop back directly on the Pi pins then?  In the Molex plug it is top row, the two on the right.

**Reply by heffneil** [reply]
Yes i tested pin 8 and 10 on the pi header with the hat off.  I wasn't sure which pins to test on molex.  I will go and jump those next and test.

**Reply by heffneil** [reply]
> You did the loop back directly on the Pi pins then? In the Molex plug it is top row, the two on the right.

When I look at the instructions, I see the pin out pic on the hat and its 1-2-3-4 where 3 and 4 are the generator send and receive.  But when I look at the cable that is connecting the mobile link I think it would plug in to pin 1 and 2?  I am wondering if the signal is on 1 and 2 and I need to move to 3 and 4?

**Reply by heffneil** [reply]
https://imgur.com/a/tyxaMGm

Here is the existing mobile link plug and the diagram of the pin out.  If I am reading it correctly I think we are in the wrong positions? How did the other guy connect everything.

**Reply by skipfire** [reply]
What other guy?

The signal on the HAT is on Pins 3 & 4, the top row 2 on the right.

![image](https://github.com/user-attachments/assets/11286787-19b3-4c30-a4a3-fa176654b46d)

**Reply by heffneil** [reply]
There was a link to another guy with a similar situation to mine.  It was linked above.  Sorry for being unclear. 

You see the pins on the mobile link are on the other 2 positions on the top row?

**Reply by skipfire** [reply]
That's at the controller end, as I said before the cable mirrors, so 1234 becomes 4321.  I'm the guy at PintSize that makes the HATs, I assure you the data pins for RS232 are the top row with the 2 pins on the right.

**Reply by heffneil** [reply]
I see the cable does flip entirely.  I have no clue what is going on here then.

**Reply by skipfire** [reply]
Have you tried the loopback on the HAT yet to ensure that the HAT is working correctly?  The HAT design does work with your controller, I've had many purchasers for Evo LCs and I don't know of any that did not get it working.

**Reply by heffneil** [reply]
I am special at screwing stuff up :)

**Reply by heffneil** [reply]
I just tested the hat.  The 2 wires are brown and green on the end of the hat.  I did not test the other end of the cable because i have to take everything apart in order to do that.

**Reply by skipfire** [reply]
And the serial test passed when testing at the HAT?  If so, then it's cable or controller.  If you are using an unmodified Molex cable between the HAT and the controller then it is very doubtful that the cable is the problem.

**Reply by heffneil** [reply]
When I loop the 2 lines you suggest (they have the brown and green cable. Here is the pic of the wiring harness (which I am sure you know)

https://imgur.com/a/7KSjLLE

Anyway the brown and green do not appear to land in 387 388 so I don't know but it doesn't seem right.

**Reply by heffneil** [reply]
> And the serial test passed when testing at the HAT? If so, then it's cable or controller. If you are using an unmodified Molex cable between the HAT and the controller then it is very doubtful that the cable is the problem.

Agreed the cable isn't bad.  What could be wrong with my controller?  I had mobilelink working so I don't know?

**Reply by skipfire** [reply]
And to confirm, you are currently testing with an unmodified Molex cable between the HAT and the controller?

**Reply by heffneil** [reply]
yes unmodified cable.  I am nearly positive that the pins you are expecting data from is not the ones you are going to get it from.  Aren't the pins for the mobile link the 2 we need to use with your hat?  If that's true I am in the blue and white plugs on the controller side but you are in the brown and green.  Please double check me because I could be backwards and upside down.  Thats why I posted those two pics?

**Reply by heffneil** [reply]
Ok I don't think the cable is wrong.  I think its the right pins.  Im lost.

**Reply by skipfire** [reply]
The HAT and cabling is correct.  I took a picture of mine, and oriented the Evo end similar to how you had yours oriented in the pic.  With the green and brown wires go from the server end where your MobileLink has the 2 populated pins to the Molex port on the HAT where I indicated the signal wires are, the top right of the plug.

If MobileLink is still working and the HAT is not when you are using the unmodified cable and no splitter, I'm out of ideas.  I could not find your order, but if you contact me through the PintSize.Me website or the facebook page I can privately discuss next steps with you.

![20240825_134605](https://github.com/user-attachments/assets/6375e8aa-a218-4ec1-b4a2-b337a7d43118)

**Comment by jgyates** [comment, marked as answer]
So all of the loopback test passed, correct?

if your generator is a Liquid Cooled model and the engine is a 4.5L engine then one thing to keep in mind is [this item form the know issues wiki entry](https://github.com/jgyates/genmon/wiki/Appendix-D-Known-Issues#18-liquid-cooled-45l-model-serial-data-rate).

You may have to chagne the data rate on the 4.5L models

**Reply by heffneil** [reply]
Yup this fixed it within 1 second.  Totally crazy! THANKS! I mean what a PITA

**Reply by jgyates** [reply]
Yeah, not sure what that one type of controller has a different data rate than the rest. I am glad it is working for you now.
